### PR TITLE
feat(DesktopViewer): introduce DesktopViewer component

### PR DESCRIPTION
**Pull request by mareklibra**
**What**:  introduce DesktopViewer component

**Link to Storybook**: https://mareklibra.github.io/patternfly-react/

**Additional issues**:

**Additional info**:
The component unifies composition of various console components
for server/virtual machine remote access (recently via serial
console or VNC).

The components unifies access to virtual machine/server console
connection via desktop (means non-browser) viewer application.

A button for downloading .vv file (of "application/x-virt-viewer" mime-type)
is rendered.
Console details are farther rendered in textual form to enable copy&paste
to a generic desktop viewer application.

This component implements common use case for any virtual machine/server
management application which provides remote console access.
First adopters of this component will be kubevirt/web-ui, ovirt-web-ui or Cockpit.

Examples of use:
```
 - <DesktopViewer spice={spice} vnc={vnc} />
 - <DesktopViewer spice={spice} vnc={vnc}>
     <div>My custom content of more-info section</div>
   </DesktopViewer>
 - <DesktopViewer spice={spice}
                  onDownload={(fileName, content, mimeType) => {/* perform saveAs action */}}
                  onGenerate={({ console, type } => { return ({ content, fileName, mimeType });})} />
```

### CODE REVIEW

**Comment by mareklibra** [comment]
Once merged, this will be added to #562

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+}
+MoreInformation.propTypes = {
+  children: PropTypes.any.isRequired
+};
+MoreInformation.defaultProps = {};
+
+const ConnectWithRemoteViewer = ({ children, onGenerate, onDownload, spice, vnc }) => {
+  const onClick = () => {
+    const console = spice || vnc; // strictly prefer spice over vnc
+    const type = spice ? 'spice' : 'vnc';
```
Comment (ohadlevy) [inline packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: use constants?

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (console) {
+      const vv = onGenerate({ console, type });
+      onDownload(vv.fileName || 'console.vv', vv.content, vv.mimeType || 'application/x-virt-viewer');
+    }
+  };
+
+  const isConnection = spice || vnc;
```
Comment (ohadlevy) [inline packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: you already defined it as ```console```?

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  return (
+    <div className="remote-viewer">
+      <div className="remote-viewer-launch">
+        <button onClick={onClick} disabled={!isConnection}>
```
Comment (ohadlevy) [inline packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: use pf-button?

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import MoreInformationContent from './MoreInformationContent';
+import { generateVVFile, downloadFile } from './vvFileGenerator';
+import consoleDetailPropType from './consoleDetailPropType';
+
+class MoreInformation extends React.Component {
```
Comment (ohadlevy) [inline packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: move this to another file? maybe it should even exists in pf?

File: packages/patternfly-3/react-console/src/DesktopViewer/DesktopViewer.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import ManualConnection from './ManualConnection';
+import ConnectWithRemoteViewer from './ConnectWithRemoteViewer';
+import consoleDetailPropType from './consoleDetailPropType';
+
+const DesktopViewer = ({ children, spice, vnc, onGenerate, onDownload, ...props }) => (
+  <Grid fluid className="desktop-viewer">
+    <Row>
+      <Col md={6}>
+        <ConnectWithRemoteViewer spice={spice} vnc={vnc} onGenerate={onGenerate} onDownload={onDownload}>
```
Comment (ohadlevy) [inline packages/patternfly-3/react-console/src/DesktopViewer/DesktopViewer.js]: can't we just detect the type here instead of checking everywhere on inner components? e.g. lets just provide a ```type=``` and render a child component based on the actual type passed?

File: packages/patternfly-3/react-console/src/DesktopViewer/vvFileGenerator.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import 'blob-polyfill';
```
Comment (ohadlevy) [inline packages/patternfly-3/react-console/src/DesktopViewer/vvFileGenerator.js]: why is this required?

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js (hunk 1 above)
Comment (mareklibra) [reply packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: done. Newly using same constants as in #562.
So far copy&paste since #562 is not yet merged. I will refactor asap.

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js (hunk 2 above)
Comment (mareklibra) [reply packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: right, thanks

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js (hunk 3 above)
Comment (mareklibra) [reply packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: done

File: packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js (hunk 4 above)
Comment (mareklibra) [reply packages/patternfly-3/react-console/src/DesktopViewer/ConnectWithRemoteViewer.js]: Done - moved to a separate file.
I'll open PR adding this component to pf-react and if accepted, I'll refactor here.

File: packages/patternfly-3/react-console/src/DesktopViewer/vvFileGenerator.js (hunk 6 above)
Comment (mareklibra) [reply packages/patternfly-3/react-console/src/DesktopViewer/vvFileGenerator.js]: Mostly due to IE: https://developer.mozilla.org/en-US/docs/Web/API/Blob#Browser_compatibility

File: packages/patternfly-3/react-console/src/DesktopViewer/DesktopViewer.js (hunk 5 above)
Comment (mareklibra) [reply packages/patternfly-3/react-console/src/DesktopViewer/DesktopViewer.js]: We need to pass both types down for `ManualConnection`.

For the `ConnectWithRemoteViewer`, the decision is just in the onClick button's handler.
The "spice over vnc" preference is hardcoded there, but I consider it as a part of the pattern.

This straightforward listing might become long if adding new types but recently just RDP is under consideration.

Since we probably need to pass the type back in the `onGenerate()` call-back, I don't see simplification in doing that - still same (2) props.

**Comment by mareklibra** [comment]
Comments above addressed. Rebased.

**Comment by mareklibra** [comment]
The `MoreInformation` component moved to pf-react 3: #602 .
Once/If merged, it's use here will be refactored.

**Comment by mareklibra** [comment]
Same issue with unintended PR closing as in #562. Happend when deploying storybook to feature branch.
I'll open new PR referencing this one.

Does anyone know how to properly deploy and re-deploy storybook to a feature branch when the PR is already open?

Branch in my fork used for this PR: `mareklibra:desktopViewer`
Storybook deployed via: `yarn storybook:build && yarn storybook:deploy --branch=desktopViewer`

**Comment by jeff-phillips-18** [comment]
@mareklibra use a branch name different from the actual branch you are using or it overwrites your branch. I typically add `-storybook` to the end fo the branch name so:
```
yarn storybook:build && yarn storybook:deploy --branch=desktopViewer-storybook
```
